src/profile-logic/process-profile.js

```diff
@@ -450,66 +563,60 @@ function adjustMarkerTimestamps(markers, delta) {
  * Throws an exception if it encounters an incompatible profile.
- * For a description of the processed format, look at docs/gecko-profile-format or
- * alternately the tests for this function.
- * @param {object} profile A profile object, in the Gecko format.
- * @return {object} A new profile object, in the 'processed' format.
+ * For a description of the processed format, look at docs/gecko-profile-format.md
  */
-export function processProfile(profile) {
+export function processProfile(geckoProfile: GeckoProfile): Profile {
   // Handle profiles from older versions of Gecko. This call might throw an
   // exception.
-  upgradeGeckoProfileToCurrentVersion(profile);
+  upgradeGeckoProfileToCurrentVersion(geckoProfile);
 
-  const libs = profile.libs;
+  const libs = geckoProfile.libs;
   let threads = [];
   const tasktracer = getEmptyTaskTracerData();
 
-  if ('tasktracer' in profile && 'threads' in profile.tasktracer) {
-    addProcessedTaskTracerData(
-      profile.tasktracer,
+  if (geckoProfile.tasktracer && geckoProfile.tasktracer.threads) {
+    _addProcessedTaskTracerData(
+      geckoProfile.tasktracer,
       tasktracer,
       libs,
-      profile.meta.startTime
+      geckoProfile.meta.startTime
     );
   }
 
-  for (const thread of profile.threads) {
-    threads.push(processThread(thread, libs));
+  for (const thread of geckoProfile.threads) {
+    threads.push(_processThread(thread, libs));
   }
 
-  for (const subprocessProfile of profile.processes) {
+  for (const subprocessProfile of geckoProfile.processes) {
     const subprocessLibs = subprocessProfile.libs;
     const adjustTimestampsBy =
-      subprocessProfile.meta.startTime - profile.meta.startTime;
+      subprocessProfile.meta.startTime - geckoProfile.meta.startTime;
     threads = threads.concat(
       subprocessProfile.threads.map(thread => {
-        const newThread = processThread(thread, subprocessLibs);
-        newThread.samples = adjustSampleTimestamps(
+        const newThread = _processThread(thread, subprocessLibs);
+        newThread.samples = _adjustSampleTimestamps(
           newThread.samples,
           adjustTimestampsBy
         );
-        newThread.markers = adjustMarkerTimestamps(
+        newThread.markers = _adjustMarkerTimestamps(
           newThread.markers,
           adjustTimestampsBy
         );
         return newThread;
       })
     );
-    if (
-      'tasktracer' in subprocessProfile &&
-      'threads' in subprocessProfile.tasktracer
-    ) {
-      addProcessedTaskTracerData(
+    if (subprocessProfile.tasktracer && subprocessProfile.tasktracer.threads) {
+      _addProcessedTaskTracerData(
         subprocessProfile.tasktracer,
         tasktracer,
         subprocessLibs,
-        profile.meta.startTime
+        geckoProfile.meta.startTime
       );
     }
   }
 
   const result = {
-    meta: Object.assign({}, profile.meta, {
+    meta: Object.assign({}, geckoProfile.meta, {
```

Comment:
Hmm... I didn't move everything over to `immutableUpdate`, only the `Object.assign`s that flow had trouble with.